crates/sui-graphql-rpc/src/server/version.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use axum::{
+    http::{HeaderMap, HeaderValue, Request, StatusCode},
+    middleware::Next,
+    response::{IntoResponse, Response},
+};
+
+use crate::error::{code, graphql_error};
+
+const VERSION_HEADER: &str = "X-Sui-RPC-Version";
+
+const RPC_VERSION_FULL: &str = env!("CARGO_PKG_VERSION");
```

Comment:
I'm not sure how I feel about using the pkg version as the "rpc" version, I know we do this in other places but I don't think this is correct and I think these are two separate things. The api that is supported by this endpoint is one version and the package version is important if we ever publish things to cratesio and should represent the version of the public api of the crate itself (the functions and traits that are exported and public from this library)

Reply:
I do agree that the decision to couple the RPC version with the workspace version doesn't make sense -- its version is different from the node version, and the SDK version so there is a lot of untangling to do there, but I don't have so strong a negative reaction to coupling the RPC version with *a* package version (in particular the version of the package that defines the RPC service binary).

This is mainly because even if we do publish that crate to `crates.io`, its only purpose will be to build that binary, i.e. no library component, so its only public interface is the GraphQL schema, so the versioning aligns.  Not a big deal to manually set the version in this file, but I chose to then use the version in the manifest because people know to go look there, and if those two versions ever got out of whack it would be more confusing.

Reply:
There's a couple scenarios where that does break down. I suppose my bigger concern will come when we build a client and want to provide the client as a library as a part of the sdk, we will not be able to use the package version for that.

Reply:
Agreed, the client should have its own versioning scheme (i.e. its crate version should not match the RPC version), but I also don't think the client should depend on this crate.  If there is anything that exists in this crate that should be shared with a client, then we should factor that out into its own crate that they both depend on, which doesn't use the RPC version (as an internal crate, I would go with your prior suggestion of it having a pre-1.0 version that we just bump on every publish).

Having said that, I think it's unlikely that there will be such sharing.  The usual thing to share would be the types returned by the GraphQL queries, but on the server-side those data-types end up coupled with server-specific logic (which fields are computed, vs directly available from the store/resolver), and on the client side, they need to deal with the fact that the query may have only requested a subset of the fields, so the data representations end up being different on both sides.